tests/async-context.test.ts

```diff
@@ -0,0 +1,249 @@
+        return AsyncContext.wrap(() => ctx.get());
+      });
+      const secondWrap = ctx.run(second, () => {
+        return AsyncContext.wrap(() => ctx.get());
+      });
+      const actual = [firstWrap(), secondWrap(), firstWrap(), secondWrap()];
+
+      assert.deepStrictEqual(actual, [first, second, first, second]);
+    });
+  });
+});
+
+describe("async via promises", () => {
+  beforeEach(() => {
+    Promise.prototype.then = then;
+  });
+  afterEach(() => {
+    Promise.prototype.then = nativeThen;
+  });
+
+  describe("run and get", () => {
+    it("get returns current context value", async () => {
+      const ctx = new AsyncContext<Value>();
+      const expected = { id: 1 };
+
+      const actual = await ctx.run(expected, () => {
+        return Promise.resolve().then(() => ctx.get());
+      });
+
+      assert.equal(actual, expected);
+    });
+
+    it("get within nesting contexts", async () => {
+      const ctx = new AsyncContext<Value>();
+      const first = { id: 1 };
+      const second = { id: 2 };
+
+      const actual = await ctx.run(first, () => {
+        return Promise.resolve([])
+          .then((temp) => {
+            temp.push(ctx.get());
+            return temp;
+          })
+          .then((temp) => {
+            return ctx.run(second, () => {
+              return Promise.resolve().then(() => {
+                temp.push(ctx.get());
+                return temp;
+              });
+            });
+          })
+          .then((temp) => {
+            temp.push(ctx.get());
+            return temp;
+          });
+      });
+
+      assert.deepStrictEqual(actual, [first, second, first]);
+    });
+
+    it("get within nesting different contexts", async () => {
+      const a = new AsyncContext<Value>();
+      const b = new AsyncContext<Value>();
+      const first = { id: 1 };
+      const second = { id: 2 };
+
+      const actual = await a.run(first, () => {
+        return Promise.resolve([])
+          .then((temp) => {
+            temp.push(a.get(), b.get());
+            return temp;
+          })
+          .then((temp) => {
+            return b.run(second, () => {
+              return Promise.resolve().then(() => {
+                temp.push(a.get(), b.get());
+                return temp;
+              });
+            });
+          })
+          .then((temp) => {
+            temp.push(a.get(), b.get());
+            return temp;
+          });
+      });
+
+      assert.deepStrictEqual(actual, [
+        first,
+        undefined,
+        first,
+        second,
+        first,
+        undefined,
+      ]);
+    });
+
+    it("get out of order", async () => {
+      const ctx = new AsyncContext<Value>();
+      const first = { id: 1 };
+      const second = { id: 2 };
+
+      const firstRun = ctx.run(first, () => {
+        return [
+          sleep(10).then(() => ctx.get()),
+          sleep(20).then(() => ctx.get()),
+          sleep(30).then(() => ctx.get()),
```

Comment:
I think you shouldn't rely on `sleep()` resolving in the order you want.  From hard experience implementing timeouts, and timing-based tests, on `XMLHttpRequest` for Mozilla code, I believe you have no real guarantee your code will resolve in the order you intend.

I'd suggest something like this:

```typescript
type PromiseCallback<T> = () => Promise<T>;

class PriorityQueuePromises<T> 
{
  #items = new Map<number, PromiseCallback<T>>;

  add(priority: number, userCallback: PromiseCallback<T>) : PromiseCallback<T>
  {
    this.#items.set(priority, userCallback);
    return userCallback;
  }

  async run() {
    const keys = Array.from(this.#items.keys());
    keys.sort();
    const callbacks = keys.map(key => this.#items.get(key) as PromiseCallback<T>);
    
    while (callbacks.length) {
      const nextCallback = callbacks.shift() as PromiseCallback<T>;
      await nextCallback();
    }
  }
}

const queue = new PriorityQueuePromises<unknown>;

const firstRun = ctx.run(first, () => {
  return [
    queue.add(10, () => ctx.get()), 
    // ...
  ]
});
// ditto for secondRun

/*
This starts them in the order you want.  Enclosed in void for linting.
We want to fire and forget at this point: Let await Promise.all() do the real work.
*/
void(queue.run()); 

// this gets you your result in the assert order you can check.
const actual = await Promise.all(firstRun.concat(secondRun));
```

Reply:
I kinda want to avoid async promise tests entirely, they're just a nuisance to get right. And, because the "polyfill" is just to call `AsyncContext.wrap()` on a callback and invoke that callback after the current `.run(…)` finishes sync, we could just fully test `wrap()` in sync testing.